vunit/ui.py

```diff
@@ -870,6 +877,20 @@ def _main(self, post_run):
         all_ok = self._main_run(post_run)
         return all_ok
 
+    def _add_ext_pkg(self):
+        """
+        Add sources corresponding to VHPIDIRECT arrays (or their placeholders)
+        """
+        extmem = "ext_pkg-novhpi.vhd"
+        if self._project.vhpi:

```

Comment:
The name of the vhpi field is to general. You are assuming it means a VHPI implementation of read_byte/write_byte was added.

Reply:
Also I think the point where the external byte array support should be selected by the user is in the add_data_types method. That method should take a file name for an alternate implementation.

Reply:
> The name of the vhpi field is to general. You are assuming it means a VHPI implementation of read_byte/write_byte was added.

Yes, it is too general. However, I am not worried about it for now. It is just a placeholder. I think that the relevant issue here is the granularity. Where should this variable be defined? Per project? Per testbench? Per configuration?

Once we decide that, I think that the name of the option should be directly related to the new name of `ext_pkg`, since this variable is precisely used when the package is imported/added.

> Also I think the point where the external byte array support should be selected by the user is in the add_data_types method. That method should take a file name for an alternate implementation.

I could not find any method named `add_data_types`. Do you mean that it should be added? If so, wouldn't it be the same as `set_objects`?

I am not sure about what you mean with *should take a file name for an alternate implementation*. Is 'file name' a C source file? A built C object? A VHDL source file?

Regarding C sources, you commented in #462 that VUnit should not built them. Also, I cannot think of a use case where `ext_pkg-*vhpi.vhd` need to be modified by the user. That's why I think that `add_data_types` and `set_objects` might be the same indeed.

For completeness, there is a use case where the use might want to provide an alternative `ext_pkg-novhpi.vhd` file: to provide a custom model which is external to VUnit but still implemented in VHDL only. However, I think that this is out of the scope of my proposal. Certainly, I would name it 'custom model', instead of 'external model'.

Reply:
> Yes, it is too general. However, I am not worried about it for now. It is just a placeholder. I think that the relevant issue here is the granularity. Where should this variable be defined? Per project? Per testbench? Per configuration?

Since we need to compile a different package to change the behaviour it needs to be per run.py file. You cannot have different versions of packages for different test benches in the VUnit project model.

> I could not find any method named add_data_types. Do you mean that it should be added? If so, wouldn't it be the same as set_objects?

I remember now that we automatically add the data_types since they are used by the built in runner. In we should make the byte array an optional component that is added explicitly by the user in the run.py file like `add_byte_array(external_implementation="pkg.vhd")`. The verification components would also add the byte array without an external implementation automatically if not already added by the user previously.

Reply:
> Since we need to compile a different package to change the behaviour it needs to be per run.py file. You cannot have different versions of packages for different test benches in the VUnit project model.

Oh, that's unfortunate. But thanks for the clarification. I did need it. Is this a constraint that applies to any package (even those written by the user) or is it only about VUnit's internal packages?

Nonetheless, I think that the solution here is to i) allow the user to provide 'objects' per configuration and ii) provide a 'dummy' C file where stubs of all the required C functions are declared. This will allow the user to set `dummy` for all the configs, and then overwrite the implementation for specific testbenches.

> I remember now that we automatically add the data_types since they are used by the built in runner. In we should make the byte array an optional component that is added explicitly by the user in the run.py file like `add_byte_array(external_implementation="pkg.vhd")`. The verification components would also add the byte array without an external implementation automatically if not already added by the user previously.

Related to a comment above, should we let the option be `add_external_arrays`, so it is possible to decide about `byte_vector`, `integer_vector` and/or `memory`?